File: PythonAlgorithmInterView/선형자료구조/두수의합.py

# rock solution brutefoce 방식으로 문제 해결 시간 복잡도는 n제곱이 걸린다.
def twoSum(self, nums: list[int], target: int) -> list[int]:

    for i in range(len(nums)):
        for j in range(i+1, len(nums)):
            if nums[i] + nums[j] == target :
                return [i, j]

# 상길북에서 보여준 n제곱이지만 첫번째 얻은 값을 빼고 in 조건으로 값 파악을 하면 ms의 차이는 확연히 들어난다.
def twoSum2(self, nums: list[int], target: int) -> list[int]:
    for i, n  in enumerate(nums):

        element = target - n

        if element in nums[ i+1:]:
            return [i, nums[i+1:].index(element)+ i + 1]

# 시간 복잡도를 생각하여 n의 시간 복잡도만 가지게 변경할 수 있다. object를 활용하여 데이터 추가.
def twoSum3(self, nums: list[int], target: int) -> list[int]:

    nums_map = {}

    # index의 값을 알아야하기 때문에 value에 index값을 넣어준다.
    for i, n  in enumerate(nums):
        nums_map[n] = i

    print(nums_map)

    for i, n in enumerate(nums):
        if target - n in nums_map and i != nums_map[target - n] :
            return [i, nums_map[target - n]]



if __name__ == '__main__':
    # nums = [3,2,4]
    # target = 6
    nums = [2,5,5,11]
    target = 10
    print(twoSum3(None, nums, target))

File: PythonAlgorithmInterView/선형자료구조/두수의합.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
 var twoSum = function(nums, target) {
    
    for(let i = 0; i < nums.length; i++){
        for(let j = i+1; j < nums.length; j++){
            if(nums[i] + nums[j] == target){
                return [i, j]
            }
        }
    }
};

// 자바스크립트 경우 이렇게 푸는게 더 느리게 된다. 왜냐하면 slice, findIndex가 n시간만큼 걸리기 때문이다.
var twoSum2 = function(nums, target) {
    
    for(let i = 0; i < nums.length; i++){
        
        let ele = target - nums[i];

        let findIndex = nums.slice(i+1).findIndex((num)=> num == ele );

        console.log(
            nums.slice(i+1), findIndex
        )

        if( findIndex >= 0 ){
            return [i, findIndex + ( i + 1)]
        }
    }
};

// 객체에 값을 할당하여 시간복잡도를 줄일 수 있다. 
var twoSum3 = function(nums, target) {
    
    let nums_map = {}

    nums.forEach(( num, index )=> nums_map[num] = index)

    for(let i = 0; i < nums.length; i++){
        console.log(nums.indexOf(target - nums[i]))
        if( nums.indexOf(target - nums[i]) >= 0  && i != nums_map[target - nums[i]] ){
            return [ i, nums_map[target - nums[i]]]
        }
    }
};

// 객체에 값을 할당하여 시간복잡도를 줄일 수 있다. 자바스크립트같은 경우 n번 만큼 걸리는 시간을 줄이는게 제일 좋다.
var twoSum4 = function(nums, target) {
    
    let nums_map = {}

    for( let i = 0; i < nums.length; i++){
        // console.log(nums_map, target - nums[i])
        if( nums_map[target - nums[i]] >= 0){
            return [nums_map[target - nums[i]], i]
        }
        nums_map[nums[i]] = i
    }
    
};
let nums = [2,7,11,15];
let target = 9
console.log(twoSum4(nums, target))